File: core/src/nl/t64/game/rpg/subjects/CollisionSubject.java
package nl.t64.game.rpg.subjects;

import com.badlogic.gdx.math.Rectangle;
import nl.t64.game.rpg.screens.world.entity.Direction;

import java.util.ArrayList;
import java.util.List;


public class CollisionSubject {

    private final List<CollisionObserver> observers = new ArrayList<>();

    public void addObserver(CollisionObserver observer) {
        observers.add(observer);
    }

    public void removeObserver(CollisionObserver observer) {
        observers.remove(observer);
    }

    public void removeAllObservers() {
        observers.clear();
    }

    public void notifyCollision(Rectangle playerBoundingBox, Direction playerDirection) {
        List.copyOf(observers).forEach(observer -> observer.onNotifyCollision(playerBoundingBox, playerDirection));
    }

}

File: core/src/nl/t64/game/rpg/subjects/CollisionSubject.kt
package nl.t64.game.rpg.subjects

import com.badlogic.gdx.math.Rectangle
import nl.t64.game.rpg.screens.world.entity.Direction


class CollisionSubject {

    private val observers: MutableList<CollisionObserver> = ArrayList()

    fun addObserver(observer: CollisionObserver) {
        observers.add(observer)
    }

    fun removeObserver(observer: CollisionObserver) {
        observers.remove(observer)
    }

    fun removeAllObservers() {
        observers.clear()
    }

    fun notifyCollision(playerBoundingBox: Rectangle, playerDirection: Direction) {
        ArrayList(observers).forEach { it.onNotifyCollision(playerBoundingBox, playerDirection) }
    }

}
